### ISSUE TITLE
original cellphonedb vs. omnipath cellphonedb not the same

### ISSUE THREAD

**Issue by giovp** [saezlab/pypath#171]
**How did you access the data?**
Python client

**Why do you think the data is wrong?**
We tested against the original database (cellphonedb) and the results seem to be different.

**An example of the wrong data**
A full notebook where the two database are tested can be found here: https://github.com/theislab/squidpy_reproducibility/pull/10

**What solution you can expect?**

Hi @deeenes and team, we'd like to report an incosistency with the cellphone database that we query from squidpy (via cellphonedb implementation of @michalk8 ) with https://github.com/saezlab/omnipath .
Basically, the reported ligan-redceptor annotation list with a test dataset was not matching between the omnipath query and the original cellphonedb database.
The issue was originally reported by a squidpy user here: https://github.com/theislab/squidpy/issues/308 . The cellphonedb database version of the latest cellphonedb software version in pypi is `2.0.0`. Is that also what's provided by omnipath? 
Thank you !

**Comment by deeenes** [comment]
Hi @giovp,

I checked the notebook, it's not completely clear for me how I could find interactions which are in CellPhoneDB but not in OmniPath, or the other way around, are in OmniPath but missing from CellPhoneDB. Maybe it would be easier for you to do so? One or a few examples are enough.

p.s. This reviewnb is really cool, but me as a guest I can't run code there, right?

Best,

Denes

**Comment by michalk8** [comment]
I've extracted 3 csv files, 1 from omnipath from the Python interface using:
```python
import omnipath as op
df = op.interactions.import_intercell_network(
    interactions_params={'resources': "CellPhoneDB"},
    transmitter_params={'categories': 'ligand'},
    receiver_params={'categories': 'receptor'}
)
```
1 from the R interface (just a sanity check - it matches [there are tests for this the `omnipath` repo, but still]) and
the last one from CellPhoneDB by saving `interactions` [here](https://github.com/Teichlab/cellphonedb/blob/3e813a86ec995e69fd938c034bcb1c13b7b7f4a1/cellphonedb/src/core/methods/method_launcher.py#L56).

[omnipath_cpdb.csv.txt](https://github.com/saezlab/pypath/files/6198583/omnipath_cpdb.csv.txt)
[omnipath_r_cpdb.csv.txt](https://github.com/saezlab/pypath/files/6198709/omnipath_r_cpdb.csv.txt)
[cpdb.csv.txt](https://github.com/saezlab/pypath/files/6198584/cpdb.csv.txt)

I haven't yet looked into the possible mismatch, though I'd just share these here for now.

**Comment by deeenes** [comment]
Hi All,

Sorry about returning to you so late, I needed to check a number of things.

First, let me tell about two known and legitimate reasons why the result of the OmniPath queries might be different from CellPhoneDB data.

1) For the ligand and receptor annotations we apply blacklists across all resources removing certain proteins which seem to be not ligands and not receptors. Because CellPhoneDB supposed to be expert curated, probably no or very few records affected, and those might be mistakes in the blacklists that we curated for OmniPath.

```
import omnipath
res_cpdb = {'resources': 'CellPhoneDB', 'entity_types': 'protein'}
icn_cpdb = omnipath.interactions.import_intercell_network(
    interactions_params = res_cpdb,
    transmitter_params = res_cpdb,
    receiver_params = res_cpdb,
)
```

2) If we query only the interactions, all ligand-receptor pairs will be included. However CellPhoneDB also contains ligand-ligand and receptor-receptor interactions. These are processed as undirected, however other resources might assign directions. Therefore we can not be sure that the output of this query contains only ligand-receptor interactions. Also unless you add the `directed = 'no'` argument the interactions without any direction information won't be included.

```
import omnipath
cpdb_interactions = (
    omnipath.interactions.PostTranslational.get(resources = 'CellPhoneDB')
)
```

Next, about the format of CellPhoneDB. The file https://github.com/Teichlab/cellphonedb/blob/master/cellphonedb/src/core/data/interaction_input.csv contains the interacting pairs in alphabetic order, i.e. `name_1` and `name_2` does not mean that the interaction is `name_1 -> name_2` or `name_1` is the ligand and `name_2` is the receptor. In `pypath` we process https://github.com/Teichlab/cellphonedb/blob/master/cellphonedb/src/core/data/protein_input.csv to find out which proteins are ligands or receptors. The receptors are labelled as receptors, the ligands are not labelled directly, we define ligands as secreted proteins or peripheral or transmembrane proteins which are not receptors. Then we swap the direction for receptor -> ligand interactions from the interaction file. While investigating your issue, I found an error in this part of the processing which I fixed and I updated the data in the web service. This issue affected around 200 interactions. With this correction still  a few interactions might fall out, the ones which are undirected (ligand-ligand or receptor-receptor), have no references and no direction from other resources. I counted these interactions, 60 of the CellPhoneDB interactions are missing due to the reason I've just mentioned, and further 4 are missing because one of the partners couldn't be identified as ligand or receptor:

```
import re
import pandas as pd
import omnipath
from pypath.inputs import cellphonedb

reup = re.compile(
    r'[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]'
    r'([A-Z][A-Z0-9]{2}[0-9]){1,2}'
)

def only_uniprots(pairs):
    return {
        i for i in pairs if
        all(isinstance(u, str) and reup.match(u) for u in i)
    }

def both_directions(ia):
    return {tuple(reversed(i)) for i in ia} | ia

cpdb_pypath = list(cellphonedb.cellphonedb_interactions())
from_pypath = {(r.id_a, r.id_b) for r in cpdb_pypath}
from_pypath = only_uniprots(from_pypath)

cpdb_webserv = omnipath.interactions.PostTranslational.get(
    resources = 'CellPhoneDB',
    directed = 'no',
)
from_webserv = {(r.source, r.target) for r in cpdb_webserv.itertuples()}
from_webserv = only_uniprots(from_webserv)

missing = {
    r for r in cpdb_pypath
    if (r.id_a, r.id_b) in (from_cpdb - from_webserv)
}

len(missing)
# 64

missing
# {CellphonedbInteraction(id_a='D3W0D1', id_b='Q6UVW9', sources='CellPhoneDB', references='24223577', interaction_type='receptor-receptor', type_a='receptor', type_b='receptor'),
#  CellphonedbInteraction(id_a='O00182', id_b='Q14118', sources='InnateDB-All;CellPhoneDB', references='', interaction_type='ligand-ligand', type_a='ligand', type_b='ligand'),
#  CellphonedbInteraction(id_a='O00182', id_b='Q92673', sources='InnateDB-All;CellPhoneDB', references='', interaction_type='ligand-ligand', type_a='ligand', type_b='ligand'),
# ...

[r for r in missing_pypath_records if r.type_a != r.type_b]
# [CellphonedbInteraction(id_a='P25024', id_b='P54577', sources='Guide2Pharma;CellPhoneDB', references='', interaction_type='receptor-unknown', type_a='receptor', type_b='unknown'),
#  CellphonedbInteraction(id_a='Q8N126', id_b='Q9H4G0', sources='CellPhoneDB', references='', interaction_type='ligand-unknown', type_a='ligand', type_b='unknown'),
#  CellphonedbInteraction(id_a='O14939', id_b='P84077', sources='Guide2Pharma;CellPhoneDB', references='', interaction_type='unknown-ligand', type_a='unknown', type_b='ligand'),
#  CellphonedbInteraction(id_a='P21730', id_b='P39019', sources='Guide2Pharma;CellPhoneDB', references='', interaction_type='receptor-unknown', type_a='receptor', type_b='unknown')]

cpdb_original = pd.read_csv('issue171/cpdb.csv.txt')
from_cpdb = {(r.name_1, r.name_2) for r in cpdb_original.itertuples()}
from_cpdb = only_uniprots(from_cpdb)

# The interactions imported by pypath are not identical to the ones read from the CPDB file
# because pypath set the directions to ligand -> receptor:
len(from_cpdb - from_pypath)
# 256

# Ignoring the directions, we find no interaction in the original CellPhoneDB data
# which is not returned by pypath.inputs.cellphonedb.cellphonedb_interactions:
from_pypath = both_directions(from_pypath)
len(from_cpdb - from_pypath)
# 0
```

Ps. I ignored complexes here just for sake of simplicity, those should behave a similar way:

```
grep omplex cpdb.csv.txt | wc -l
# 322
curl 'https://omnipathdb.org/interactions?resources=CellPhoneDB&entity_types=complex&directed=no' | head -n -1 | wc -l
# 322
```

Best,

Denes

**Comment by giovp** [comment]
Hi @deeenes ,

thanks a lot for the very detailed reply! and thanks @michalk8 for providing the files!
So just to recap for my understanding, inconsistencies might arise because of the way the cellphonedb annotation is "casted" into omnipath standards (in terms of data structure and processing). These are:
- filtering black-listed annotations
- definition of interactions as ligand-receptors (excluding e.g. ligands-ligands interactions)
- definition of ligands (transmembrane, secreted, not receptors)

is this correct?

Beside the filtering/processing/annotations choices, this is a very useful reference for users on how the two resources are interfaced.

**Comment by deeenes** [comment]
Hi @giovp,

Yes, all correct.

> Hi @deeenes ,
> 
> thanks a lot for the very detailed reply! and thanks @michalk8 for providing the files!
> So just to recap for my understanding, inconsistencies might arise because of the way the cellphonedb annotation is "casted" into omnipath standards (in terms of data structure and processing). These are:
> 
>     * filtering black-listed annotations

This is the reason why we can not recover the full list of CPDB ligands and receptors from OmniPath, although at CPDB this effect I guess is minor.

> 
>     * definition of interactions as ligand-receptors (excluding e.g. ligands-ligands interactions)

At least in the `import_intercell_network` query. Actually, if you pass `transmitter_param = {'causality': 'both'}, receiver_param = {'causality': 'both'}` to this function, also the ligand-ligand and receptor-receptor will be included.

> 
>     * definition of ligands (transmembrane, secreted, not receptors)

This is more of an issue of the CPDB file format, you can see how it looks like: https://github.com/Teichlab/cellphonedb/blob/master/cellphonedb/src/core/data/protein_input.csv

> 
> 
> is this correct?
> 
> Beside the filtering/processing/annotations choices, this is a very useful reference for users on how the two resources are interfaced.

Agree :)

**Comment by giovp** [comment]
sorry for late reply, thanks again a lot @deeenes for the clarification on this, will close the issue.
